rpc_util.go

```diff
@@ -874,14 +874,35 @@ func decompress(compressor encoding.Compressor, d mem.BufferSlice, maxReceiveMes
 		return nil, 0, err
 	}
 
-	out, err := mem.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1), pool)
+	out, err := mem.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)), pool)
 	if err != nil {
 		out.Free()
 		return nil, 0, err
 	}
+	if err = checkReceiveMessageOverflow(int64(out.Len()), int64(maxReceiveMessageSize), dcReader); err != nil {
+		return nil, out.Len() + 1, err
```

Comment:
Adding 1 to `out.Len()` can result in an overflow on 32 bit systems. The  following part of the function seems strange to me: 
> Optionally, if data will be over maxReceiveMessageSize, just return the size

I suggest making the following change to avoid this:
1. Declare a global error for indicating that the max receive size is exceeded:
```go
var	errMaxMessageSizeExceeded = errors.New("max message size exceeded")
```  
2. When the check here fails, `nil, 0, errMaxMessageSizeExceeded`. Update the godoc to mention the same.
3. In the caller, i.e. recvAndDecompress, instead of checking `if size > maxReceiveMessageSize`, check `if err == errMaxMessageSizeExceeded`. Also update the error message to not mention the actual size, because we didn't read the entire message anyways: `grpc: received message after decompression larger than max %d`.

This ensures we're using the returned error to indicate the failure instead of relying on special values of the bytes read count.

Reply:
wait, in case of overflow i think we are still reading the partial data so we need to return the length upto which we have read?

Reply:
so, the logic should be read till maxMessageSize -1 first and then check if we still have bytes to read by reading one byte and if we overflow, throw the error along with lenght of how much we have read